lib/Sema/CSDiagnostics.cpp

```diff
@@ -101,6 +101,52 @@ Expr *FailureDiagnostic::getArgumentExprFor(Expr *anchor) const {
   return nullptr;
 }
 
+ConstraintLocator *
+FailureDiagnostic::getConstraintLocatorFor(Expr *expr) const {
+  auto locator = CS.getConstraintLocator(expr);
+
+  if (isa<DeclRefExpr>(expr)) {
+    locator = CS.getConstraintLocator(expr);
+  } else if (isa<UnresolvedDeclRefExpr>(expr)) {
+    locator = CS.getConstraintLocator(expr);
+  } else if (isa<SubscriptExpr>(expr)) {
+    locator = CS.getConstraintLocator(expr, ConstraintLocator::SubscriptMember);
+  } else if (isa<UnresolvedDotExpr>(expr)) {
+    auto UDE = dyn_cast<UnresolvedDotExpr>(expr);
+    if (isa<ConstructorRefCallExpr>(UDE->getBase())) {
+      locator =
+          CS.getConstraintLocator(expr, ConstraintLocator::ConstructorMember);
+    } else {
+      locator = CS.getConstraintLocator(expr, ConstraintLocator::Member);
+    }
+  } else if (isa<MemberRefExpr>(expr)) {
+    auto MRE = dyn_cast<MemberRefExpr>(expr);
+    if (isa<ConstructorRefCallExpr>(MRE->getBase())) {
+      locator =
+          CS.getConstraintLocator(expr, ConstraintLocator::ConstructorMember);
+    } else {
+      locator = CS.getConstraintLocator(expr, ConstraintLocator::Member);
+    }
+  } else if (isa<CallExpr>(expr)) {
+    auto CE = cast<CallExpr>(expr);
+    auto Fn = CE->getFn();
+    locator = CS.getConstraintLocator(Fn, ConstraintLocator::FunctionResult);
+  } else if (isa<ApplyExpr>(expr)) {
+    auto AE = cast<ApplyExpr>(expr);
+    auto Fn = AE->getFn();
+    locator = CS.getConstraintLocator(Fn, ConstraintLocator::ApplyFunction);
+  } else if (isa<UnresolvedMemberExpr>(expr)) {
+    locator = CS.getConstraintLocator(expr, ConstraintLocator::MemberRefBase);
```

Comment:
The type variable choice is going to be bound to has a `UnresolvedMember` because it's a result passed to `addUnresolvedMemberConstraint` in CSGen